src/iter/checked_product.rs

```diff
@@ -0,0 +1,95 @@
+use crate::identities::One;
+use crate::CheckedMul;
+
+/// This trait represents types of which an iterator can be multiplied up with overflow checking.
+/// This trait should rarely be called directly.
+pub trait CheckedProduct<Result = Self> {
+    /// Multiplies up the elements of an iterator, returning `None` if an overflow would occur.
+    ///
+    /// Multiplies up an empty iterator returns a value representing One.
+    ///
+    /// If the iterator contains Zero, the order of elements may effect whether the result is `None`.
+    fn checked_product<I: Iterator<Item = Self>>(iter: I) -> Option<Result>;
+}
+
+impl<T> CheckedProduct<T> for T
+where
+    T: CheckedMul + One,
+{
+    fn checked_product<I: Iterator<Item = Self>>(mut iter: I) -> Option<Self> {
+        iter.try_fold(Self::one(), |acc, x| acc.checked_mul(&x))
+    }
+}
+
+impl<'a, T> CheckedProduct<T> for &'a T
+where
+    T: CheckedMul + Sized + One,
+{
+    fn checked_product<I: Iterator<Item = Self>>(mut iter: I) -> Option<T> {
+        iter.try_fold(T::one(), |acc, x| acc.checked_mul(x))
+    }
+}
+
+///This trait is for iterators that can be multiplied up with overflow checking.
+trait CheckedProductIter<T, Result>: Iterator<Item = T> {
+    /// Multiplies up the elements of an iterator, returning `None` if an overflow would occur.
+    ///
+    /// Multiplies up an empty iterator returns a value representing One.
+    ///
+    /// If the iterator contains Zero, the order of elements may effect whether the result is `None`.
+    fn checked_product(self) -> Option<Result>;
```

Comment:
Similarly, for consistency with `Iterator`, this trait should have no parameters, and the method should be parameterized `checked_product<P: CheckedProduct<Self::Item>>`.

I think then we could also combine your two iterator extensions into one `CheckedIterator` trait, or even `NumIterator` to leave room for more extension methods.

Reply:
> Rather than a new `iter` module here, I think it would make more sense to move this to the `num-iter` crate.

I wasn't sure. I'm very happy to close this pull request and move the code changes over there if that's best.